### ISSUE TITLE
tf.test.compute_gradient (v2) expects wrong empty gradient shape

### ISSUE THREAD

**Issue by duncanriach** [tensorflow/tensorflow#33660]
**System Information**

- Have I written custom code: YES (see below)
- OS Platform and Distribution: Ubuntu 18.04.3 LTS (inside `tensorflow/tensorflow:2.0.0-gpu`)
- Mobile device: N/A
- TensorFlow installed from: binary (inside `tensorflow/tensorflow:2.0.0-gpu`)
- TensorFlow version: 2.0.0
- Python version: 2.7.15+
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: 10.0
- GPU model and memory: TITANV 12GB

**Current Behavior**
When the gradient tensors are totally empty (i.e. all their dimensions are zero), and more than one input is considered at the same time,  the shape of an expected gradient tensor does not match the actual shape produced by the op's gradient function. This case is not currently tested in `tensorflow/python/ops/gradient_checker_v2_test.py` (see `testEmptyMatMul`).

*UPDATE*: I also tried the configuration in `testEmptyMatMul` (see additional code below) and found that this also tickles the apparent bug. Surprisingly, `//tensorflow/python:gradient_checker_v2_test` passes when running on tag v2.0.0; I don't currently understand why.

I have tested
  * with two different ops, each with two inputs (both fail).
  * cases where not all of the gradient tensors are empty (as in `testEmptyMatMul`).
  * `tf.compat.v1.test.compute_gradient` with more than one empty input does **not** exhibit this behavior, so it was probably not present in `tensorflow/python/ops/gradient_checker.py` but was introduced in `tensorflow/python/ops/gradient_checker_v2.py`. See code at the end, in the "TensorFlow Version 1 API" section, that demonstrates this.

I have **not** tested
  * using ops with more than two inputs

**Expected Behavior**
These exceptions should not be produced when using built-in ops. Either the ops' gradient functions are producing the wrong shaped tensors (which would be a bug), or, more likely, the code in `gradient_checker_v2.py` is expecting the wrong shape.

**Code to reproduce the issue**
```
import numpy as np
import tensorflow as tf

def empty(rank):
  shape = (0,) * rank
  return np.array([]).reshape(shape)

# comment-out the first to run the second
tf.test.compute_gradient(tf.nn.bias_add, [empty(3), empty(1)])
tf.test.compute_gradient(tf.linalg.matmul, [empty(2), empty(3)])
```

The following is essentially the same as the code in `testEmptyMatMul` in `tensorflow/python/ops/gradient_checker_v2_test.py`: 
```
import numpy as np
import tensorflow as tf

def random_tensor(shape):
  return tf.constant(np.random.random_sample(shape))

def f(x, y):
  return tf.linalg.matmul(x, y)

x = random_tensor((0, 3))
y = random_tensor((3, 4))

jacobians = tf.test.compute_gradient(f, [x, y])
```

**Other info / logs**

Output from the above code:
```
Traceback (most recent call last):
  File "repro_eager_issue.py", line 24, in <module>
    tf.test.compute_gradient(tf.nn.bias_add, [empty(3), empty(1)])
  File "/usr/local/lib/python2.7/dist-packages/tensorflow_core/python/ops/gradient_checker_v2.py", line 332, in compute_gradient
    return _compute_gradient_list(f, x, delta)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow_core/python/ops/gradient_checker_v2.py", line 293, in _compute_gradient_list
    xs, i, delta) for i in range(len(xs))])
  File "/usr/local/lib/python2.7/dist-packages/tensorflow_core/python/ops/gradient_checker_v2.py", line 278, in _compute_gradient
    xs, param)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow_core/python/ops/gradient_checker_v2.py", line 187, in _compute_theoretical_jacobian
    (x.shape, grad.shape))
ValueError: Empty gradient has wrong shape: expected (0,), got (0, 0, 0)

Traceback (most recent call last):
  File "repro_eager_issue.py", line 25, in <module>
    tf.test.compute_gradient(tf.linalg.matmul, [empty(2), empty(3)])
  File "/usr/local/lib/python2.7/dist-packages/tensorflow_core/python/ops/gradient_checker_v2.py", line 332, in compute_gradient
    return _compute_gradient_list(f, x, delta)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow_core/python/ops/gradient_checker_v2.py", line 293, in _compute_gradient_list
    xs, i, delta) for i in range(len(xs))])
  File "/usr/local/lib/python2.7/dist-packages/tensorflow_core/python/ops/gradient_checker_v2.py", line 278, in _compute_gradient
    xs, param)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow_core/python/ops/gradient_checker_v2.py", line 187, in _compute_theoretical_jacobian
    (x.shape, grad.shape))
ValueError: Empty gradient has wrong shape: expected (0, 0, 0), got (0, 0)

Traceback (most recent call last):
  File "tf_issue_33660.py", line 77, in <module>
    existing_test_repro()
  File "tf_issue_33660.py", line 71, in existing_test_repro
    jacobians = tf.test.compute_gradient(f, [x, y])
  File "/usr/local/lib/python2.7/dist-packages/tensorflow_core/python/ops/gradient_checker_v2.py", line 332, in compute_gradient
    return _compute_gradient_list(f, x, delta)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow_core/python/ops/gradient_checker_v2.py", line 293, in _compute_gradient_list
    xs, i, delta) for i in range(len(xs))])
  File "/usr/local/lib/python2.7/dist-packages/tensorflow_core/python/ops/gradient_checker_v2.py", line 278, in _compute_gradient
    xs, param)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow_core/python/ops/gradient_checker_v2.py", line 187, in _compute_theoretical_jacobian
    (x.shape, grad.shape))
ValueError: Empty gradient has wrong shape: expected (3, 4), got (0, 3)
```

**Work-Around**

`compute_gradient` can be called multiple times, once for each input, to get the analytical and numerical jacobians for each input separately. As, follows:

```
import numpy as np
import tensorflow as tf

def empty(rank):
  shape = (0,) * rank
  return np.array([]).reshape(shape)

input_val = empty(3)
bias_val = empty(1)

def bias_add_1(input_val):
  return tf.nn.bias_add(input_val, bias_val)

def bias_add_2(bias_val):
  return tf.nn.bias_add(input_val, bias_val)

input_jacobians = tf.test.compute_gradient(bias_add_1, [input_val])
bias_jacobians = tf.test.compute_gradient(bias_add_2, [bias_val])
```

**TensorFlow Version 1 API**

The following code **does not** throw an exception, demonstrating that this problem does not exist with `tf.compat.v1.test.compute_gradient`.

```
import numpy as np
import tensorflow as tf

def empty_tensor(shape):
  return tf.constant([], shape=shape)

tf.compat.v1.disable_eager_execution()
input_shape = output_shape = (0, 0, 0)
bias_shape = (0,)
input_tensor = empty_tensor(input_shape)
bias_tensor = empty_tensor(bias_shape)
output_tensor = tf.nn.bias_add(input_tensor, bias_tensor)
with tf.compat.v1.Session() as sess:
  jacobians = tf.compat.v1.test.compute_gradient(
      [input_tensor, bias_tensor], [input_shape, bias_shape], output_tensor,
      output_shape)
```

**Comment by oanush** [comment]
Issue replicating for TF-2.0, kindly find the [gist](https://colab.sandbox.google.com/gist/oanush/839b2999d53195a5dd8c98f22c31ab50/33660.ipynb) of the colab.Thanks!

**Comment by duncanriach** [comment]
I just added an update to the description: "I also tried the configuration in `testEmptyMatMul` (see additional code) and found that this also tickles the apparent bug. This suggests that `tensorflow/python/ops/gradient_checker_v2_test.py` is not being run as part of the CI. Perhaps `gradient_checker_test.py` and `gradient_checker_v2_test.py` should be moved to `tensorflow/python/kernel_tests` and included in BUILD.

I also added some code to the end of the description that demonstrates that `tf.compat.v1.test.compute_gradient` does not exhibit this behavior.

**Comment by duncanriach** [comment]
Surprisingly, `//tensorflow/python:gradient_checker_v2_test` just passed on master on 2019-11-01. I'm currently running it on tag:v2.0.0.

**Comment by gadagashwini** [comment]
@duncanriach, I tried with Tf 2.1, its working without any error.
Please find the gist [here](https://colab.sandbox.google.com/gist/gadagashwini/0c11957c6fca4f94461cd57b63040389/untitled456.ipynb) and please close if issue is resolved. Thanks

**Comment by duncanriach** [comment]
Thank you, @gadagashwini. I can confirm that this issue is resolved in TensorFlow version 2.1.0. Closing.
